src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,303 @@
-#endif
-  if (!stream.is_open())
+  bool success = OpenFile(filename, opts, true, stream);
+  if (!success)
   {
     Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "'. " << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "'; load failed."
-          << std::endl;
-
     return false;
   }
 
-  FileType loadType = inputLoadType;
-  std::string stringType;
-  if (inputLoadType == FileType::AutoDetect)
-  {
-    // Attempt to auto-detect the type from the given file.
-    loadType = AutoDetect(stream, filename);
-    // Provide error if we don't know the type.
-    if (loadType == FileType::FileTypeUnknown)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << "Unable to detect type of '" << filename << "'; "
-            << "incorrect extension?" << std::endl;
-      else
-        Log::Warn << "Unable to detect type of '" << filename << "'; load "
-            << "failed. Incorrect extension?" << std::endl;
-
-      return false;
-    }
-  }
-
-  stringType = GetStringType(loadType);
-
-#ifndef ARMA_USE_HDF5
-  if (inputLoadType == FileType::HDF5Binary)
+  success = DetectFileType<MatType>(filename, opts, true, &stream);
+  if (!success)
   {
-    // Ensure that HDF5 is supported.
     Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Attempted to load '" << filename << "' as HDF5 data, but "
-          << "Armadillo was compiled without HDF5 support.  Load failed."
-          << std::endl;
-    else
-      Log::Warn << "Attempted to load '" << filename << "' as HDF5 data, but "
-          << "Armadillo was compiled without HDF5 support.  Load failed."
-          << std::endl;
-
     return false;
   }
-#endif
-
-  // Try to load the file; but if it's raw_binary, it could be a problem.
-  if (loadType == FileType::RawBinary)
-    Log::Warn << "Loading '" << filename << "' as " << stringType << "; "
-        << "but this may not be the actual filetype!" << std::endl;
-  else
-    Log::Info << "Loading '" << filename << "' as " << stringType << ".  "
-        << std::flush;
 
-  // We can't use the stream if the type is HDF5.
-  bool success;
-  LoadCSV loader;
-
-  if (loadType != FileType::HDF5Binary)
+  if constexpr (IsArma<MatType>::value || IsSparseMat<MatType>::value)
   {
-    if (loadType == FileType::CSVASCII)
-      success = loader.LoadNumericCSV(matrix, stream);
+    TextOptions txtOpts(std::move(opts));
+    success = LoadMatrix(filename, matrix, stream, txtOpts);
+    opts = std::move(txtOpts);
+  }
+  else 
+  {
+    if (opts.Fatal())
+      Log::Fatal << "DataOptionType is unknown!."
+        "please use a known type or provide specific overloads" << std::endl;
     else
-      success = matrix.load(stream, ToArmaFileType(loadType));
+      Log::Warn << "DataOptionType is unknown!."
+          "please use a known type or provide specific overloads" << std::endl;
   }
```

Comment:
```suggestion

    return false;
  }
```

Double-check me on this one, but I think we need to make sure that loading fails here.